src/huggingface_hub/hf_api.py

```diff
@@ -1826,6 +1834,11 @@ def upload_file(
             path_or_fileobj = os.path.normpath(os.path.expanduser(path_or_fileobj))
             if not os.path.isfile(path_or_fileobj):
                 raise ValueError(f"Provided path: '{path_or_fileobj}' is not a file")
+            if os.stat(path_or_fileobj).st_size // 1_000_000_000 >= 30:
+                raise ValueError(
+                    f"The file {path_or_fileobj} is larger than 30GB and cannot be uploaded "
+                    "with `upload_file`. Please use the `Repository` object instead."
```

Comment:
Should we offer any additional guidance here? As a user, you may find yourself sort of out of luck if you've already planned to use the mixin for this (if that's where it ends up being surfaced)

Reply:
Will be discussed in a separate issue https://github.com/huggingface/huggingface_hub/issues/995.